Pull request for CSV adapter.

sakagg opened the pull request:
I have initiated a pull request so that it becomes easier for you to review changes made by me. I also removed the extra files and synced with upstream before initiating the PR.

davidbau commented on .gitignore:
```diff
@@ -7,3 +7,4 @@ example/*.js
 example/*.map
 test/*.js
 test/*.map
+TEST_2G
```
What's test_2g?

sakagg replied:
That's just a folder of for some parsers I was trying. It contains various CSS and HTML parsers I was looking at while making my proposal. I'll remove that in the next commit.

davidbau commented on src/controller.coffee:
```diff
@@ -1853,13 +1878,13 @@ define ['droplet-helper',
       unless @textFocus.handwritten
         newParse = null
-        string = @textFocus.stringify(@mode.empty).trim()
+        string = @mode.normalString @textFocus.stringify(@mode.empty)
```
might we want all the logic from line 1881 to 1890 to be part of the base Parser normalString, i.e., trim and look at quotes?  Then specific modes can override to have their whole own idea of what quotes are etc.

davidbau commented on src/controller.coffee:
```diff
@@ -4003,18 +4035,17 @@ define ['droplet-helper',
     if line of @lineNumberTags
       lineDiv = @lineNumberTags[line]
-
+      
```
search and remove trailing spaces on lines of your code.

davidbau commented on src/controller.coffee:
```diff
@@ -958,6 +961,28 @@ define ['droplet-helper',
+      lineNumber = @findLineNumberAtCoordinate(mainPoint.y)
+
+      if hitTestBlock.addButtonRect? and hitTestBlock.addButtonRect.contains mainPoint
+        console.log "Plus on: ", lineNumber
```
remove console.log

davidbau commented on src/view.coffee:
```diff
@@ -1642,12 +1646,38 @@ define ['droplet-helper', 'droplet-draw', 'droplet-model'], (helper, draw, model
+        ctx.fillStyle = '#000'
+
+        if 'add-button' in @model.classes
+          @addButtonRect.stroke(ctx, '#000')
```
instead of stroking the rect directly, try getting a path (via rect.toPath) and then using the path's beveling feature to draw a beveled path.  Can you try it to see how it looks?

sakagg commented:
Hey
 - I've fixed the above mentioned issues.
 - Added tests for CSV Adapter, but I am still looking for a way t test button.
 - As far as bevels are concerned, I am getting strange errors on setting the bevel variable to true so I have removed that part for now. I'll fix them and update the repo.

davidbau commented on src/controller.coffee:
```diff
@@ -1948,6 +1968,15 @@ define ['droplet-helper',
+    if newBlock?.start?.next?
+      block.start.append newBlock.start.next
+      newBlock.end.prev.append block.end
+      block.notifyChange()
```
It's interesting that you replace the contents of the block but do not replace the block itself, which is counterintuitive.  Was there a reason that you needed to do that?  Also, I'd love Anthony to weigh in on whether anything is missing here.

sakagg replied:
I did that intentionally. This is because if the users removes everything from the block, the block itself should not disappear, only the contents should. In other words I can say that the buttons should have the capability to modify the contents of the block, not the block itself (or it's structure).
Here is a pictorial representation.
Original Block
![original](https://cloud.githubusercontent.com/assets/6389325/7104748/33740806-e114-11e4-871a-aeeb63c8bd76.png)
If contents are replaced
![replace_contents](https://cloud.githubusercontent.com/assets/6389325/7104749/422fe464-e114-11e4-9beb-b0c6b0e30e3c.png)
If entire block is replaced
![replace_block](https://cloud.githubusercontent.com/assets/6389325/7104750/4ba1114e-e114-11e4-9184-ccea6e0089bf.png)


On the other hand, we can have cases when the block should also be modified (eg: hide the - button if no children left in block) in which case we should replace the entire block. This will require modifying the parser such that is parses an empty string to something meaningful.

davidbau commented on src/controller.coffee:
```diff
@@ -958,6 +961,28 @@ define ['droplet-helper',
+
+      if hitTestBlock.addButtonRect? and hitTestBlock.addButtonRect.contains mainPoint
+        #console.log "Plus on: ", lineNumber
+        line = @mode.handleButton @getValue().split('\n')[lineNumber], 'add-button'
```
handleButton should not work at the line-level, but at the block level.

You should probably pass handleButton three arguments
(1) the entire text for the block (which may be multiline)
(2) possibly the set of classes for the block (so that it's easy to tell which type of block it is)
(3) the command - 'add-button'

David

sakagg replied:
I felt the same and suggested it few days back on the google group (https://groups.google.com/forum/#!msg/pencilcode-development/kcdiFv6GtdA/mTYMcqyX9EUJ), but since I didn't get an affirmation, I didn't work on it. Now that you have confirmed it, I went ahead and implemented it and you can see it in the latest commit.

Saksham

davidbau commented on src/view.coffee:
```diff
@@ -1642,12 +1646,46 @@ define ['droplet-helper', 'droplet-draw', 'droplet-model'], (helper, draw, model
         for size, i in @minDimensions
           size.width = Math.max size.width,
               @view.opts.tabWidth + @view.opts.tabOffset
+          size.width += @extraWidth
```
For a multiline block, you're adding @extrawidth to every line.  The +/- button should be on the last line, so should just apply there.

davidbau commented on src/view.coffee:
```diff
@@ -1642,12 +1646,46 @@ define ['droplet-helper', 'droplet-draw', 'droplet-model'], (helper, draw, model
+      drawSelf: (ctx, style) ->
+        super
+
+        start = @totalBounds.x + @totalBounds.width - @extraWidth
```
Need to think through positioning in the multiline case.  Currently you put the +/- buttons in the upper-right corner of the bounding rect for the block.  We want it to be on the last line.  But lower-right corner isn't right: you should figure the actual width of the last line to position the +/-.
